dnssec-tools/validator/doc/getaddr.1
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GETADDR 1"
.TH GETADDR 1 "2011-06-30" "perl v5.12.4" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
getaddr \- command\-line test program for the val_getaddrinfo() function
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    getaddr [options] <hostname|IPv4 address|IPv6 address>
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This utility is a command-line wrapper around the \fI\fIval_getaddrinfo()\fI\fR function. 
It invokes the \fI\fIval_getaddrinfo()\fI\fR operation for the given command-line
arguments and displays the return code from the function and the contents of the
\&\fIaddrinfo\fR structure.
.PP
The exit status for this program is 1 if the answer received is trusted, 2 if 
the answer received is validated, and \-1 for an error. If no validation is performed,
or the result is neither trusted nor validated, a value of 0 is returned. The trusted and
validated status values are obtained using the \fI\fIval_istrusted()\fI\fR and
\&\fI\fIval_isvalidated()\fI\fR functions from \fI\fIlibval\fI\|(3)\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-c, \-\-canonname" 4
.IX Item "-c, --canonname"
Use the \s-1AI_CANONNAME\s0 flag in \fI\fIval_getaddrinfo()\fI\fR.
.IP "\-s, \-\-service=\fIPORT|SERVICE\fR" 4
.IX Item "-s, --service=PORT|SERVICE"
Use the specified transport-layer port or service name.
.IP "\-o, \-\-output=<debug\-level>:<dest\-type>[:<dest\-options>]" 4
.IX Item "-o, --output=<debug-level>:<dest-type>[:<dest-options>]"
<debug\-level> is 1\-7, corresponding to syslog levels ALERT-DEBUG
<dest\-type> is one of file, net, syslog, stderr, stdout
<dest\-options> depends on <dest\-type>
    file:<file\-name>   (opened in append mode)
    net[:<host\-name>:<host\-port>] (127.0.0.1:1053
    syslog[:facility] (0\-23 (default 1 \s-1USER\s0))
.IP "\-h, \-\-help" 4
.IX Item "-h, --help"
Display the help and exit.
.SH "PRE-REQUISITES"
.IX Header "PRE-REQUISITES"
libval
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2005\-2011 \s-1SPARTA\s0, Inc.  All rights reserved.
See the \s-1COPYING\s0 file included with the DNSSEC-Tools package for details.
.SH "AUTHORS"
.IX Header "AUTHORS"
Abhijit Hayatnagarkar, Suresh Krishnaswamy, Robert Story
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fB\f(BIval_getaddrinfo\fB\|(3)\fR
.PP
\&\fB\f(BIlibval\fB\|(3)\fR
.PP
http://dnssec\-tools.sourceforge.net


dnssec-tools/NEWS
This file contains a list of major changes per release.  See the
ChangeLog file for a complete set of changes and their details
details.


1.10
 - New Features:
    - New Apps:     (see the validator/apps directory for details)
                    - dnssec-check: check dnssec support from your ISP
                    - dnssec-nodes: graphically displays a DNS
		      hierarchy, color coded by each node's DNSSEC status
                    - dnssec-system-tray: displays pop-up
     		      notifications when a libval-enabled application
		      triggers a DNSSEC error
                    - lookup: a graphical DNS lookup utility that
		      displays the results in a hierarchical tree and
		      color codes the window according to DNSSEC status

    - libval:       - Added support for building on Windows.
                    - added support for falling back to recursion when
                      the caching name server does not appear to 
                      support DNSSEC. This also works as a mechanism
                      to work around poisoned or misbehaving cache. 
                    - Significant improvements to the the asynchronous support. 
    - lsdnssec:     - Improvements to lsdnssec to display different
                      output depending on whether a zone is a
                      stand-alone zone or under control of rollerd.
    - nagios:       - Plugins for the nagios monitoring system which
                      enable monitoring of zone rollerover states.
    - firefox:      - Improved patches that work with the most recent firefox

 Plus many more minor features and bug fixes

1.9
 - New Features:
    - lsdnssec:     - Added a new flag (-p) to show only zones in a
                      particular rollerd phase.
                    - fixed bugs to align timing output with rollerd.
    - rollerd:      - Added a -logtz flag for logging timezones
                    - fixed bugs related to the -alwayssign flag.
		    - zonesigner's path is taken from the config file.
    - rollctl:      - Added -rollall and -rollzone options.
    - zonesigner:   - Assumes keys need to be generated for new zones
		      (Assumes -genkeys option was given if a keyrec file
		      can't be found.)
                    - Exits with unique exit codes if a failure occurs.
                      ("zonesigner -xc CODE" can lookup a description for it.")
		    - Added the -phase option so rollover options could be
		      more easily specified.
    - lights:       - A simple GUI to check the status of rollover states
    - blinkenlights:- Added hide/show commands for rollrec names and zone
		      names, for split-zone support
    - cleankrf:	    - Fixed deletion of obsolete set keyrecs.
    - GUI commands: - Fixed how the Exit command works so they don't coredump.

    - libsres
      & libval:     - New beta support for issuing asynchronous requests.
                      This can speed up queries by up to 4 times if used.
                      (see example code in validator/apps/validator_selftest.c)
                    - NSEC3, DLV and IPv6 are enabled by default.
                    - improved logging and logging-callback support.
    - drawvalmap    - Can output PNG files now

 - Packaging:
                    - Our download page now allows you to download
                      the C validator libraries independently of the
                      full DNNSEC-Tools tool-suite.

 - Many bugs were also fixed in the 240+ changes.

1.8
 - New Features:
    - zonesigner, rollerd
                    - Made changes so that these tools are more compatible 
                      with recent versions of Bind
                    - The zone_errors configuration parameter allows a zone-
                      specific maximum to be set.  Once exceeded, that zone
                      will be skipped rather than allowing rollover to continue.
    - blinkenlights
                    - Recognizes when rollerd abruptly quits, so error messages
                      aren't spewed interminably.
    - ZonFile::Fast - Fixed parsing of DS records containing spaces and 
                      parsing of mname and rname SOA fields
                    - Added support for parsing KEY records
    - keyrec.pm     - Made changes to properly lock keyrec files before 
                      writing to them.
                    - Begun process of deprecating keyrec_open().
    - mapper:       - added a new option: --node-size for mapping
                      complex zones.
    - dnspktflow:   - added two new options:
                      --layout-style for selecting the layout style to use
                      --node-size for mapping complex zones.
                    - Add new (default) option to cluster
                      authoritative nodes together to help better
                      understand the relationships between traffic
                      patterns and authoritative name server/zone arrangement.
    - libval:       - Now distributed with the Root TA.
                    - Added stricter checks for openssl SHA-256 support in 
                      configure. 
                    - Added several improvements that allow the validator to 
                      lookup information within provably insecure zones that 
                      do not handle EDNS0 requests nicely. This includes 
                      adding support for turning off EDNS0 when traversing a 
                      name hierarchy that leads to a provably insecure zone, 
                      EDNS0 fallback support, and additional checks to check 
                      the sanity of response data.
                    - Fixed certain bugs in CNAME handling and in the 
                      validation of proofs accompanying wildcard responses, 
                      referrals and alias chains.
                    - Fixed support for RSADSA and RSASHA-512 signature 
                      validation.

dnssec-tools/validator/apps/getaddr.c
/*
 * Copyright 2005-2011 SPARTA, Inc.  All rights reserved.
 * See the COPYING file distributed with this software for details.
 *
 * Author: Abhijit Hayatnagarkar
 *
 * A command-line tool for testing the val_getaddrinfo() function.
 */
#include "validator-config.h"
#include <validator/validator.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define	NAME	"getaddr"
#define	VERS	"version: 1.0"
#define	DTVERS	"DNSSEC-Tools version: 1.8"

#ifdef HAVE_GETOPT_LONG
// Program options
static struct option prog_options[] = {
    {"help", 0, 0, 'h'},
    {"canonname", 0, 0, 'c'},
    {"service", 0, 0, 's'},
    {"Version", 0, 0, 'V'},
    {0, 0, 0, 0}
};
#endif

void
usage(char *progname)
{
    fprintf(stderr,
            "Usage: %s [options] <hostname|IPv4 address|IPv6 address>\n",
            progname);
    fprintf(stderr, "Options:\n");
    fprintf(stderr,
            "\t-h, --help                      display usage and exit\n");
    fprintf(stderr,
            "\t-c, --canonname                 use the AI_CANONNAME flag\n");
    fprintf(stderr,
            "\t-s, --service=<PORT|SERVICE>    transport-layer port or service name\n");
    fprintf(stderr,
            "\t-o, --output=<debug-level>:<dest-type>[:<dest-options>]\n"
            "\t          <debug-level> is 1-7, corresponding to syslog levels\n"
            "\t          <dest-type> is one of file, net, syslog, stderr, stdout\n"
            "\t          <dest-options> depends on <dest-type>\n"
            "\t              file:<file-name>   (opened in append mode)\n" 
            "\t              net[:<host-name>:<host-port>] (127.0.0.1:1053\n" 
            "\t              syslog[:facility] (0-23 (default 1 USER))\n" );
    fprintf(stderr,
            "\t-V, --Version                   display version and exit\n");

}

void
version(void)
{
     fprintf(stderr, "%s %s\n", NAME, VERS);
     fprintf(stderr, "%s\n", DTVERS);
}

#define ADDRINFO_TYPE     0
#define VAL_ADDRINFO_TYPE 1

static void
print_addrinfo(int type, void *ainfo)
{
    struct sockaddr_in *s_inaddr = NULL;
    struct sockaddr_in6 *s_in6addr = NULL;
    struct addrinfo *a = (struct addrinfo *) ainfo;
    char            buf[INET6_ADDRSTRLEN];
    size_t          buflen = INET6_ADDRSTRLEN;

    while (a != NULL) {
        printf("{\n");
        printf("\tFlags:     %d [", a->ai_flags);
        if (a->ai_flags & AI_PASSIVE)
            printf("AI_PASSIVE ");
        if (a->ai_flags & AI_CANONNAME)
            printf("AI_CANONNAME ");
        if (a->ai_flags & AI_NUMERICHOST)
            printf("AI_NUMERICHOST ");
#ifdef AI_V4MAPPED
        if (a->ai_flags & AI_V4MAPPED)
            printf("AI_V4MAPPED ");
#endif
#ifdef AI_ALL
        if (a->ai_flags & AI_ALL)
            printf("AI_ALL ");
#endif
#ifdef AI_ADDRCONFIG
        if (a->ai_flags & AI_ADDRCONFIG)
            printf("AI_ADDRCONFIG ");
#endif
        //              if (a->ai_flags & AI_NUMERICSERV) printf("AI_NUMERICSERV ");
        printf("]\n");
        printf("\tFamily:    %d [%s]\n", a->ai_family,
               (a->ai_family == AF_UNSPEC) ? "AF_UNSPEC" :
               (a->ai_family == AF_INET) ? "AF_INET" :
               (a->ai_family == AF_INET6) ? "AF_INET6" : "Unknown");
        printf("\tSockType:  %d [%s]\n", a->ai_socktype,
               (a->ai_socktype == SOCK_STREAM) ? "SOCK_STREAM" :
               (a->ai_socktype == SOCK_DGRAM) ? "SOCK_DGRAM" :
               (a->ai_socktype == SOCK_RAW) ? "SOCK_RAW" : "Unknown");
        printf("\tProtocol:  %d [%s]\n", a->ai_protocol,
               (a->ai_protocol == IPPROTO_IP) ? "IPPROTO_IP" :
               (a->ai_protocol == IPPROTO_TCP) ? "IPPROTO_TCP" :
               (a->ai_protocol == IPPROTO_UDP) ? "IPPROTO_UDP" :
               "Unknown");
        printf("\tAddrLen:   %d\n", a->ai_addrlen);

        if (a->ai_addr != NULL) {
            const char *addr = NULL;
            printf("\tAddrPtr:   %p\n", a->ai_addr);
            if (a->ai_family == AF_INET) {
                s_inaddr = (struct sockaddr_in *) (a->ai_addr);
                INET_NTOP(AF_INET, ((struct sockaddr *)s_inaddr), sizeof(s_inaddr),
                          buf, buflen, addr);
                printf("\tAddr:      %s\n", addr);
            } else if (a->ai_family == AF_INET6) {
                s_in6addr = (struct sockaddr_in6 *) (a->ai_addr);
                INET_NTOP(AF_INET6, ((struct sockaddr *)s_in6addr), sizeof(s_in6addr),
                          buf, buflen, addr);
                printf("\tAddr:      %s\n", addr);
            } else
                printf
                    ("\tAddr:      Cannot parse address. Unknown protocol family\n");
        } else
            printf("\tAddr:      (null)\n");

        if (a->ai_canonname)
            printf("\tCanonName: %s\n", a->ai_canonname);
        else
            printf("\tCanonName: (null)\n");

        printf("}\n");

        a = (struct addrinfo *) (a->ai_next);
    }
}


int
main(int argc, char *argv[])
{
    char           *node = NULL;
    char           *service = NULL;
    struct addrinfo hints;
    struct addrinfo *val_ainfo = NULL;
    int             retval;
    int             getcanonname = 0;
    int             portspecified = 0;
    val_log_t      *logp;
    val_status_t val_status;

    // Parse the command line
    while (1) {
        int             c;
#ifdef HAVE_GETOPT_LONG
        int             opt_index = 0;
#ifdef HAVE_GETOPT_LONG_ONLY
        c = getopt_long_only(argc, argv, "hco:s:V",
                             prog_options, &opt_index);
#else
        c = getopt_long(argc, argv, "hco:s:V", prog_options, &opt_index);
#endif
#else                           /* only have getopt */
        c = getopt(argc, argv, "hco:s:V");
#endif

        if (c == -1) {
            break;
        }

        switch (c) {
        case 'h':
            usage(argv[0]);
            return -1;
        case 'o':
            logp = val_log_add_optarg(optarg, 1);
            if (NULL == logp) { /* err msg already logged */
                usage(argv[0]);
                return -1;
            }
            break;

        case 's':
            portspecified = 1;
            service = optarg;
            break;
        case 'c':
            getcanonname = 1;
            break;
        case 'V':
            version();
            return 0;
        default:
            fprintf(stderr, "Invalid option %s\n", argv[optind - 1]);
            usage(argv[0]);
            return -1;
        }
    }

    if (optind < argc) {
        node = argv[optind++];
    } else {
        fprintf(stderr, "Error: node name not specified\n");
        usage(argv[0]);
        return -1;
    }

    memset(&hints, 0, sizeof(struct addrinfo));
    if (getcanonname) {
        hints.ai_flags |= AI_CANONNAME;
    }

    {
        retval = val_getaddrinfo(NULL, node, service, &hints, &val_ainfo, &val_status);
        printf("Return code = %d\n", retval);
        printf("Validator status code = %d (%s)\n", val_status, p_val_status(val_status));

        if (retval != 0) {
#ifdef HAVE_GAI_STRERROR
            printf("Error in val_getaddrinfo(): %s\n",
                   gai_strerror(retval));
#else
            printf("Error in val_getaddrinfo(): %d\n", retval);
#endif
            return -1;
        } else {
            print_addrinfo(VAL_ADDRINFO_TYPE, val_ainfo);
        }

        /*
         * cleanup 
         */
        freeaddrinfo(val_ainfo);
        if (val_isvalidated(val_status)) {
            return 2; 
        } 
        if (val_istrusted(val_status)) {
            return 1; 
        } 
        
    } 
    return 0;
}
